### PR TITLE
[Feature] Add RepVGG backbone and checkpoints

### PR DESCRIPTION
**Pull request by zhangrui-wolf**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add RepVGG backbone.

## Modification

Add the RepVGG source code, and its accompanying configuration files and related files for deployment..

## BC-breaking (Optional)

No

## Use cases (Optional)

No

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/models/backbones/repvgg.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+            conv_weight = torch.zeros((self.in_channels, input_dim, 3, 3),
+                                      dtype=branch.weight.dtype)
+            for i in range(self.in_channels):
+                conv_weight[i, i % input_dim, 1, 1] = 1
+            conv_weight = conv_weight.to(branch.weight.device)
+            running_mean = branch.running_mean
+            running_var = branch.running_var
+            gamma = branch.weight
+            beta = branch.bias
+            eps = branch.eps
```
[the remark this answers is not in the document]
Comment (zhangrui-wolf) [reply mmcls/models/backbones/repvgg.py]: Done!

File: mmcls/models/backbones/repvgg.py [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+    def reparameterize(self):
+        """Fuse the parameters in all branches of repvgg.
+
+        Returns:
+            (Weight, bias) (tuple): Parameters after fusion of all branches.
+                the first element is the weight and the second is the bias.
+        """
+        weight_3x3, bias_3x3 = self._fuse_conv_bn(self.branch_3x3)
+        weight_1x1, bias_1x1 = self._fuse_conv_bn(self.branch_1x1)
+        weight_1x1 = F.pad(weight_1x1, [1, 1, 1, 1])
```
[the remark this answers is not in the document]
Comment (zhangrui-wolf) [reply mmcls/models/backbones/repvgg.py:194]: Done!

File: mmcls/models/backbones/repvgg.py [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+            gamma = branch.norm.weight
+            beta = branch.norm.bias
+            eps = branch.norm.eps
+        else:
+            input_dim = self.in_channels // self.groups
+            conv_weight = torch.zeros((self.in_channels, input_dim, 3, 3),
+                                      dtype=branch.weight.dtype)
+            for i in range(self.in_channels):
+                conv_weight[i, i % input_dim, 1, 1] = 1
+            conv_weight = conv_weight.to(branch.weight.device)
```
[the remark this answers is not in the document]
Comment (zhangrui-wolf) [reply mmcls/models/backbones/repvgg.py]: I don't think so.

File: mmcls/models/backbones/repvgg.py [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+        dict(
+            num_blocks=[4, 6, 16, 1],
+            width_factor=[3, 3, 3, 5],
+            group_idx=g2_map),
+        'B3g4':
+        dict(
+            num_blocks=[4, 6, 16, 1],
+            width_factor=[3, 3, 3, 5],
+            group_idx=g4_map),
+        'D2se':
```
[the remark this answers is not in the document]
Comment (zhangrui-wolf) [reply mmcls/models/backbones/repvgg.py:364]: Done！

File: mmcls/models/backbones/repvgg.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+                channels, out_channels, num_blocks, stride, dilation,
+                next_create_block_idx, se_cfg, init_cfg)
+            stage_name = f'stage_{i + 1}'
+            self.add_module(stage_name, stage)
+            self.stages.append(stage_name)
+
+            channels = out_channels
+
+    def _make_stage(self, in_channels, out_channels, num_blocks, stride,
+                    dilation, next_create_block_idx, se_cfg, init_cfg):
```
[the remark this answers is not in the document]
Comment (zhangrui-wolf) [reply mmcls/models/backbones/repvgg.py]: Done!

File: mmcls/models/backbones/repvgg.py [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+        for i, stage_name in enumerate(self.stages):
+            stage = getattr(self, stage_name)
+            x = stage(x)
+            if i in self.out_indices:
+                outs.append(x)
+
+        if len(outs) == 1:
+            return outs[0]
+        else:
+            return tuple(outs)
```
[the remark this answers is not in the document]
Comment (zhangrui-wolf) [reply mmcls/models/backbones/repvgg.py:534]: Can I use it like this now?

File: mmcls/models/backbones/repvgg.py [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+            self.branch_1x1 = self.create_conv_bn(kernel_size=1)
+
+        if se_cfg is not None:
+            self.se_layer = SELayer(channels=out_channels, **se_cfg)
+        else:
+            self.se_layer = None
+
+        self.act = build_activation_layer(act_cfg)
+
+    def create_conv_bn(self, kernel_size, dilation=1, padding=0):
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:106]: Can we use `mmcv.cnn.ConvModule` to replace it?

File: configs/_base_/datasets/pipelines/auto_aug.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+sub_policy_22 = [
+    dict(type='Invert', prob=0.6),
+    dict(type='Equalize', prob=1.0)
+]
+sub_policy_23 = [
+    dict(type='ColorTransform', magnitude=4, prob=0.6),
+    dict(type='Contrast', magnitude=8, prob=1.0)
+]
+
+auto_increasing_policies = [
```
Comment (mzr1996) [inline configs/_base_/datasets/pipelines/auto_aug.py:98]: Maybe reuse autoaug policies style in [#395](https://github.com/open-mmlab/mmclassification/blob/ac0bba5ec164633c80264e589a7208ebe47a2473/configs/_base_/datasets/pipelines/auto_aug.py)?
Will these sub-policies be reused in other policies? If not, it’s not necessary to split them to multiple variables.

File: mmcls/models/backbones/repvgg.py (hunk 7 above)
Comment (Ezra-Yu) [reply mmcls/models/backbones/repvgg.py:106]: if we use `mmcv.cnn.ConvModule`, the key name of checkpoint will change and the fuction of `fuse_conv_bn` will be harder to implement。

File: configs/_base_/datasets/pipelines/auto_aug.py (hunk 8 above)
Comment (Ezra-Yu) [reply configs/_base_/datasets/pipelines/auto_aug.py:98]: OK.

File: mmcls/models/backbones/repvgg.py (hunk 7 above)
Comment (mzr1996) [reply mmcls/models/backbones/repvgg.py:106]: ok

File: configs/_base_/datasets/pipelines/auto_aug.py (hunk 8 above)
Comment (Ezra-Yu) [reply configs/_base_/datasets/pipelines/auto_aug.py]: since there is no official auto_aug in repvgg repo and training accuracy is not aligned, so use auto_aug in [#395](https://github.com/open-mmlab/mmclassification/blob/ac0bba5ec164633c80264e589a7208ebe47a2473/configs/_base_/datasets/pipelines/auto_aug.py) temporarily.

File: configs/repvgg/metafile.yml [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        - re-parameterization Convolution
+        - VGG-style Neural Network
+    Paper:
+        URL: https://arxiv.org/abs/2101.03697
+        Title: 'RepVGG: Making VGG-style ConvNets Great Again'
+    README: configs/repvgg/README.md
+
+Models:
+  - Name: repvggA0_4xb64-coslr-120e_in-1k
+    In Collection: Repvgg
```
Comment (mzr1996) [inline configs/repvgg/metafile.yml:17]: RepVGG

File: configs/repvgg/metafile.yml [hunk 10]
```diff
@@ -0,0 +1,219 @@
+Collections:
+  - Name: RepVGG
+    Metadata:
+      Training Data: ImageNet
+      Epochs: 200
```
Comment (mzr1996) [inline configs/repvgg/metafile.yml:5]: I notice that some models use 120 epochs. If `Epochs` is not a shared field of all models, please move it to each model.

File: configs/repvgg/metafile.yml [hunk 11]
```diff
@@ -0,0 +1,219 @@
+Collections:
+  - Name: RepVGG
+    Metadata:
+      Training Data: ImageNet
```
Comment (mzr1996) [inline configs/repvgg/metafile.yml:4]: ImageNet-1k

File: configs/repvgg/metafile.yml [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    README: configs/repvgg/README.md
+
+Models:
+  - Name: repvggA0_4xb64-coslr-120e_in-1k
+    In Collection: Repvgg
+    Config: configs/repvgg/repvggA0_4xb64-coslr-120e_in-1k.py
+    Metadata:
+      FLOPs: 1520000000
+      Parameters: 9110000
+      Training Data: ImageNet
```
Comment (mzr1996) [inline configs/repvgg/metafile.yml:22]: `Training Data` is shared by all models, just put it in the `RepVGG` collection. It's not necessary to put it in each model.

File: configs/repvgg/README.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
+# Repvgg: Making vgg-style convnets great again
+
+## Introduction
+
```
Comment (mzr1996) [inline configs/repvgg/README.md:4]: Please add `<!-- [ALGORITHM] -->`, and it will be used to count algorithms in docs.

File: mmcls/models/backbones/repvgg.py [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+            padding=self.padding,
+            dilation=self.dilation,
+            groups=self.groups,
+            bias=True)
+        self.branch_reparam.weight.data = reparam_weight
+        self.branch_reparam.bias.data = reparam_bias
+
+        for param in self.parameters():
+            param.detach_()
+        self.__delattr__('branch_3x3')
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:181]: please use `delattr` instead of calling the magic method directly.

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: configs/repvgg/deploy/repvggA0_4xb64-coslr-120e_in-1k_deploy.py [hunk 15]
```diff
@@ -0,0 +1,3 @@
+_base_ = '../repvggA0_4xb64-coslr-120e_in-1k.py'
+
+model = dict(backbone=dict(deploy=True))
```
[the remark this answers is not in the document]
Comment (Ezra-Yu) [reply configs/repvgg/deploy/repvggA0_4xb64-coslr-120e_in-1k_deploy.py:3]: the checkpoint name will change correspondingly?

File: mmcls/models/backbones/repvgg.py [hunk 16]
```diff
@@ -0,0 +1,531 @@
...
+                kernel_size=3,
+                stride=stride,
+                padding=padding,
+                dilation=dilation,
+                groups=groups,
+                bias=True,
+                padding_mode=padding_mode)
+        else:
+            if out_channels == in_channels and stride == 1 and \
+                    padding == dilation:
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:87]: So this condition is to judge if input shape and output shape are the same?
I think we can add a comment here.

Comment (Ezra-Yu) [reply mmcls/models/backbones/repvgg.py:89]: Done.

File: mmcls/models/backbones/repvgg.py [hunk 17]
```diff
@@ -0,0 +1,533 @@
...
+                bias_3x3 + bias_1x1 + bias_norm)
+
+    def _fuse_conv_bn(self, branch):
+        """Fuse the parameters in a branch with a conv and bn.
+
+        Args:
+            branch (Sequential): A branch with conv and bn in repvggblock.
+
+        Returns:
+            (weight, bias) (tuple): The parameters obtained after
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:216]: Same to above

File: mmcls/models/backbones/repvgg.py [hunk 18]
```diff
@@ -0,0 +1,533 @@
...
+
+    def _norm_to_conv3x3(self, branch_nrom):
+        """Convert a norm to a conv3x3-bn Sequential.refer to the paper:
+        RepVGG: Making VGG-style ConvNets Great Again for more detail
+
+        Args:
+            branch (nn.BatchNorm2d): A branch only with bn in repvggblock.
+
+        Returns:
+            tmp_conv3x3 (Sequential): a sequential with conv3x3 and bn.
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:243]: Same to above, and "Sequential" -> "Sequence".

File: mmcls/models/backbones/repvgg.py [hunk 19]
```diff
@@ -0,0 +1,533 @@
...
+            weight_norm, bias_norm = self._fuse_conv_bn(tmp_conv_bn)
+
+        return (weight_3x3 + weight_1x1 + weight_norm,
+                bias_3x3 + bias_1x1 + bias_norm)
+
+    def _fuse_conv_bn(self, branch):
+        """Fuse the parameters in a branch with a conv and bn.
+
+        Args:
+            branch (Sequential): A branch with conv and bn in repvggblock.
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:213]: "Sequential" -> "Sequence"
"repvggblock" -> "the block"

File: mmcls/models/backbones/repvgg.py [hunk 20]
```diff
@@ -0,0 +1,533 @@
...
+
+        std = (running_var + eps).sqrt()
+        fused_weight = (gamma / std).reshape(-1, 1, 1, 1) * conv_weight
+        fused_bias = -running_mean * gamma / std + beta
+
+        return fused_weight, fused_bias
+
+    def _norm_to_conv3x3(self, branch_nrom):
+        """Convert a norm to a conv3x3-bn Sequential.refer to the paper:
+        RepVGG: Making VGG-style ConvNets Great Again for more detail
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:237]:
```suggestion
        """Convert a norm layer to a conv3x3-bn sequence.
```

File: mmcls/models/backbones/repvgg.py [hunk 21]
```diff
@@ -0,0 +1,533 @@
...
+        tmp_conv3x3 = self.create_conv_bn(kernel_size=3)
+        tmp_conv3x3.conv.weight.data = conv_weight
+        tmp_conv3x3.norm = branch_nrom
+        return tmp_conv3x3
+
+
+@BACKBONES.register_module()
+class RepVGG(BaseBackbone):
+    """RepVGG backbone.
+
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:262]: Add paper name and link here. Refers to [vit backbone](https://github.com/open-mmlab/mmclassification/blob/master/mmcls/models/backbones/vision_transformer.py)

File: mmcls/models/backbones/repvgg.py [hunk 22]
```diff
@@ -0,0 +1,533 @@
...
+            with width_factor together.
+        out_indices (Sequence[int]): Output from which stages.
+        strides (Sequence[int]): Strides of the first block of each stage.
+        dilations (Sequence[int]): Dilation of each stage.
+        with_cp (bool): Use checkpoint or not. Using checkpoint will save some
+        frozen_stages (int): Stages to be frozen (all param fixed). -1 means
+            not freezing any parameters. Default: -1.
+        conv_cfg (dict | None): The config dict for conv layers. Default: None.
+        norm_cfg (dict): The config dict for norm layers.
+        act_cfg (dict): Config dict for activation layer.
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:281]: Please add default values.

File: mmcls/models/backbones/repvgg.py [hunk 23]
```diff
@@ -0,0 +1,533 @@
...
+            - group_layer_map (dict | None): RepVGG Block that declares
+                the need to apply group convolution.
+            - se_cfg (dict | None): Se Layer config
+        in_channels (int): Number of input image channels. Default: 3.
+        base_channels (int): Base channels of RepVGG backbone, work
+            with width_factor together.
+        out_indices (Sequence[int]): Output from which stages.
+        strides (Sequence[int]): Strides of the first block of each stage.
+        dilations (Sequence[int]): Dilation of each stage.
+        with_cp (bool): Use checkpoint or not. Using checkpoint will save some
```
Comment (mzr1996) [inline mmcls/models/backbones/repvgg.py:276]: Add default values.

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: mmcls/models/backbones/repvgg.py (hunk 18 above)
Comment (Ezra-Yu) [reply mmcls/models/backbones/repvgg.py:243]: in mmcv and torch, both are "Sequential", changing to "Sequence" may lead to misunderstanding

Comment (mzr1996) [reply mmcls/models/backbones/repvgg.py:243]: Sorry, I thought it's `typing.Sequence`.
Please use `nn.Module` or `nn.Sequential`

File: mmcls/models/backbones/repvgg.py (hunk 23 above)
Comment (Ezra-Yu) [reply mmcls/models/backbones/repvgg.py]: done.

File: mmcls/models/backbones/repvgg.py (hunk 21 above)
Comment (Ezra-Yu) [reply mmcls/models/backbones/repvgg.py:259]: ok

**Review by mzr1996 — APPROVED** [review]
LGTM
